### PR TITLE
v1.7 backports 2020-11-23

### PR DESCRIPTION
**Pull request by christarazi**
* #10440 -- maps/ctmap: unexport NewMap, MapType type and related consts #10440 (@tklauser)
* #11684 -- cilium: improve bpf dp signal upon ct insertion error (@borkmann)
* #13912 -- ctmap: GC orphan SNAT entries (@brb)
* #12832 -- metrics: add cilium_datapath_nat_gc_entries (@ArthurChiao)

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 10440 11684 13912 12832; do contrib/backporting/set-labels.py $pr done 1.7; done
```

### CODE REVIEW

**Comment by christarazi** [comment]
test-backport-1.7

**Comment by christarazi** [comment]
test-backport-1.7

**Comment by christarazi** [comment]
@brb Looks like https://github.com/cilium/cilium/pull/13912 requires a bit more diligence to backport. The latest compilation error I have is because of `signal.SignalWakeGC` which was introduced by https://github.com/cilium/cilium/pull/11684. Would you mind having a look into finishing this backport?

File: pkg/bpf/bpf_linux.go [hunk 1]
```diff
@@ -162,7 +162,7 @@ func LookupElementFromPointers(fd int, structPtr unsafe.Pointer, sizeOfStruct ui
 	}
 
 	if ret != 0 || err != 0 {
-		return fmt.Errorf("Unable to lookup element in map with file descriptor %d: %s", fd, err)
```
Comment (errordeveloper) [inline pkg/bpf/bpf_linux.go:165]: `%w` is fairly recent addition, does version of Go use in 1.7 support it?

Comment (errordeveloper) [reply pkg/bpf/bpf_linux.go:165]: Looks like 1.7 is on Go 1.13, so luckily it was the first version where [this feature was added](https://blog.golang.org/go1.13-errors)!

**Review by brb — APPROVED** [review]
LGTM for my changes, thanks.

**Review by brb — COMMENTED** [review]
Sure, I will take care of this backport.

**Comment by brb** [comment]
test-backport-1.7

**Comment by brb** [comment]
`	 kube-system   cilium-twv6q                   0/1     ImagePullBackOff   0          10m   192.168.36.11   k8s1   <none>` :-(

**Comment by brb** [comment]
test-backport-1.7

**Comment by brb** [comment]
test-backport-1.7

**Comment by brb** [comment]
test-backport-1.7

**Comment by brb** [comment]
test-backport-1.7

**Comment by brb** [comment]
Oh, snap, more stuff needs to be backported:
```
Stderr:
 	 # github.com/cilium/cilium/pkg/maps/ctmap [github.com/cilium/cilium/pkg/maps/ctmap.test]
	 pkg/maps/ctmap/ctmap_privileged_test.go:226:15: undefined: mapTypeIPv4AnyGlobal
	 pkg/maps/ctmap/ctmap_privileged_test.go:229:14: undefined: newMap
	 pkg/maps/ctmap/ctmap_privileged_test.go:229:35: undefined: mapTypeIPv4AnyGlobal
	 pkg/maps/ctmap/ctmap_privileged_test.go:235:15: undefined: mapTypeIPv4TCPGlobal
	 pkg/maps/ctmap/ctmap_privileged_test.go:238:14: undefined: newMap
	 pkg/maps/ctmap/ctmap_privileged_test.go:238:35: undefined: mapTypeIPv4TCPGlobal
	 pkg/maps/ctmap/ctmap_privileged_test.go:372:15: undefined: mapTypeIPv6AnyGlobal
	 pkg/maps/ctmap/ctmap_privileged_test.go:375:16: undefined: newMap
	 pkg/maps/ctmap/ctmap_privileged_test.go:375:37: undefined: mapTypeIPv6AnyGlobal
	 pkg/maps/ctmap/ctmap_privileged_test.go:381:15: undefined: mapTypeIPv6TCPGlobal
	 pkg/maps/ctmap/ctmap_privileged_test.go:381:15: too many errors
	 make: *** [tests-privileged] Error 1
```

**Comment by brb** [comment]
test-backport-1.7

**Comment by brb** [comment]
test-backport-1.7

**Pull request merged**